Multicast cannot be enabled on more than one Yggdrasil process (port reuse)

When running multiple instances in same machine, admin port should not be static. It would be nice increase by one when port is reserved instead of panic :smile:. Then should automatically change new free port to configuration. Could be temporarily fixed by editing config manually when first started.

2018/12/05 21:00:52 Starting up...
2018/12/05 21:00:52 Found 5 multicast interface(s)
2018/12/05 21:00:52 Starting switch
2018/12/05 21:00:52 Starting router
2018/12/05 21:00:52 Multicast discovery is enabled
2018/12/05 21:00:52 Failed to start multicast interface
2018/12/05 21:00:52 An error occurred during startup
panic: listen udp6 [::]:9001: bind: address already in use

goroutine 1 [running]:
main.main()
	/src/yggdrasil.go:227 +0x15bc

Correction: When changing port number from configuration manually doesn't work as expected AdminListen: tcp://localhost:9001 (usually) change to   AdminListen: tcp://localhost:9002 but still same error.
2018/12/05 21:19:31 Starting up...
2018/12/05 21:19:31 Found 5 multicast interface(s)
2018/12/05 21:19:31 Starting switch
2018/12/05 21:19:31 Starting router
2018/12/05 21:19:31 Multicast discovery is enabled
2018/12/05 21:19:31 Failed to start multicast interface
2018/12/05 21:19:31 An error occurred during startup
2018/12/05 21:19:31 Listening for TCP on: [::]:45097
panic: listen udp6 [::]:9001: bind: address already in use

goroutine 1 [running]:
main.main()
	/src/yggdrasil.go:227 +0x15bc

--- a/src/yggdrasil/multicast.go
+++ b/src/yggdrasil/multicast.go
@@ -1,6 +1,7 @@
 package yggdrasil
 
 import (
+	"context"
 	"fmt"
 	"net"
 	"time"
@@ -35,7 +36,10 @@ func (m *multicast) start() error {
 			return err
 		}
 		listenString := fmt.Sprintf("[::]:%v", addr.Port)
-		conn, err := net.ListenPacket("udp6", listenString)
+		lc := net.ListenConfig{
+			Control: multicastReuse,
+		}
+		conn, err := lc.ListenPacket(context.Background(), "udp6", listenString)
 		if err != nil {
 			return err
 		}

--- a/src/yggdrasil/multicast_other.go
+++ b/src/yggdrasil/multicast_other.go
@@ -1,4 +1,4 @@
-// +build !linux,!darwin,!netbsd,!freebsd,!openbsd,!dragonflybsd
+// +build !linux,!darwin,!netbsd,!freebsd,!openbsd,!dragonflybsd,!windows
 
 package yggdrasil
 

--- a/src/yggdrasil/multicast_other.go
+++ b/src/yggdrasil/multicast_other.go
@@ -1,0 +1,9 @@
+// +build !linux,!darwin,!netbsd,!freebsd,!openbsd,!dragonflybsd
+
+package yggdrasil
+
+import "syscall"
+
+func multicastReuse(network string, address string, c syscall.RawConn) error {
+	return nil
+}

--- a/src/yggdrasil/multicast_unix.go
+++ b/src/yggdrasil/multicast_unix.go
@@ -1,0 +1,22 @@
+// +build linux darwin netbsd freebsd openbsd dragonflybsd
+
+package yggdrasil
+
+import "syscall"
+import "golang.org/x/sys/unix"
+
+func multicastReuse(network string, address string, c syscall.RawConn) error {
+	var control error
+	var reuseport error
+
+	control = c.Control(func(fd uintptr) {
+		reuseport = unix.SetsockoptInt(int(fd), unix.SOL_SOCKET, unix.SO_REUSEPORT, 1)
+	})
+
+	switch {
+	case reuseport != nil:
+		return reuseport
+	default:
+		return control
+	}
+}

--- a/src/yggdrasil/multicast_windows.go
+++ b/src/yggdrasil/multicast_windows.go
@@ -1,0 +1,22 @@
+// +build windows
+
+package yggdrasil
+
+import "syscall"
+import "golang.org/x/sys/windows"
+
+func multicastReuse(network string, address string, c syscall.RawConn) error {
+	var control error
+	var reuseaddr error
+
+	control = c.Control(func(fd uintptr) {
+		reuseaddr = windows.SetsockoptInt(windows.Handle(fd), windows.SOL_SOCKET, windows.SO_REUSEADDR, 1)
+	})
+
+	switch {
+	case reuseaddr != nil:
+		return reuseaddr
+	default:
+		return control
+	}
+}

